### PR TITLE
ore: add enhanced panic handler

### LINKED ISSUE #24088: Logging and panics can interleave on same line in output
### What version of Materialize are you using?

73b6355108d6

### What is the issue?

In https://buildkite.com/materialize/tests/builds/71794#018c8ed6-1027-43c0-9949-c2867ff0f215 the INFO logging and panic message ended up on the same line, which I believe should not happen:
```
cluster-materialized-1  | thread 'timely:work-3' panicked at src/expr/src/scalar/func/impls/string.rscluster-u1-replica-u1: 2023-12-22T00:10:10.083219Z  INFO tokio_postgres::connection: NOTICE: relation "consensus" already exists, skipping    
cluster-materialized-1  | :1025:9:
cluster-materialized-1  | forced panic
```
It seems that the panic message doesn't write the entire line atomically, I guess having our own panic handler and synchronizing it with the logging would help: https://doc.rust-lang.org/nomicon/panic-handler.html

The immediate problem is that this breaks the parsing we do to detect known panics/errors in CI, see https://github.com/MaterializeInc/materialize/blob/main/doc/developer/ci-regexp.md
But I believe this can also cause confusion on production systems when trying to grep for a specific string and having it interleaved with others.

### PR DESCRIPTION
**Pull request by benesch**
This commit adds a method to the `ore::panic` module that installs a panic hook that enhances the message provided out of the box by the Rust runtime.

The most important enhancement is that the message is printed while holding the stderr lock, which should prevent the message from being badly interleaved with concurrent log output from the `tracing` crate. Care is taken to fall back to outputting directly to stderr in case a thread has wedged while holding the stderr lock.

A secondary enhancement is that backtraces are always collected and printed even if `RUST_BACKTRACE` has not been set. The idea is that panics are always unexpected and sometimes rare, and we don't want to miss our chance to print the cause, even if it can be a bit more verbose.

The abort-on-panic behavior of the existing `set_abort_on_panic` method in the module is also folded into the new `install_enhanced_handler` method.

Fix #24088.

### Motivation

  * This PR fixes a recognized bug.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

**Review by def- — COMMENTED** [review]
This sounds great to me, thank you! Triggered nightly: https://buildkite.com/materialize/nightly/builds/9356

**Comment by benesch** [comment]
Ugh, forgot to update a few binaries to call the new method. Just pushed up a fix for that. You may need to retrigger nightlies.

**Comment by def-** [comment]
So far so good: https://buildkite.com/materialize/nightly/builds/9357#_

File: src/ore/src/panic.rs [hunk 1]
```diff
@@ -32,37 +43,147 @@ task_local! {
...
+        // first so that we minimize the number of calls to the underlying
+        // `write(2)` system call. This minimizes the chance of (but does not
+        // outright prevent) interleaving output with C or C++ libraries that
+        // may be writing to the stderr stream outside of the Rust runtime.
+        //
+        // See https://github.com/rust-lang/rust/issues/64413 for details.
+        let mut buf = String::new();
+        let _ = writeln!(
+            &mut buf,
+            "thread '{thread_name}' panicked at {location}:\n{msg}\n{backtrace}"
```
[the remark this answers is not in the document]
Comment (benesch) [reply src/ore/src/panic.rs]: Ah, no, I refactored this a few times while writing this code and I left it in a silly state. Thanks for catching.

File: src/ore/src/panic.rs [hunk 2]
```diff
@@ -32,37 +43,147 @@ task_local! {
...
+            let buf = Arc::clone(&buf);
+            move || {
+                let mut stderr = io::stderr().lock();
+                let mut buf = buf.lock().unwrap();
+                if let Some(buf) = buf.take() {
+                    let _ = stderr.write_all(buf.as_bytes());
+                }
+
+                // Abort while still holding the stderr lock to ensure the panic
+                // is the last output printed to stder.
```
[the remark this answers is not in the document]
Comment (benesch) [reply src/ore/src/panic.rs]: 👍🏽 thanks, fixed.

File: src/ore/src/panic.rs [hunk 3]
```diff
@@ -32,37 +43,147 @@ task_local! {
...
+            // we have to string munge it ourselves. The relevant frames are
+            // between the call to `backtrace::__rust_begin_short_backtrace` and
+            // `backtrace::__rust_end_short_backtrace`, which are easy to sniff
+            // out. To make this string munging as robust as possible, if we
+            // don't find the first marker frame, we  leave the full backtrace
+            // in place.
+            let mut lines = backtrace.lines();
+            if lines
+                .find(|l| l.contains("backtrace::__rust_end_short_backtrace"))
+                .is_some()
```
[the remark this answers is not in the document]
Comment (benesch) [reply src/ore/src/panic.rs:124]: Indeed, but I find the short-circuiting behavior of `any` to be a little sketchy to rely on! At least, `find` much more obviously to me leaves the iterator at the found position.

**Comment by benesch** [comment]
Optimistically set for automerge given a green-looking nightlies and @petrosagg's approval. @def- give me a ping if this ends up causing problems in CI, or feel free to just revert.

**Comment by petrosagg** [comment]
Also, thank you for fixing this! Beyond CI auto-detecting panics this has always been frustrating for me when manually inspecting logs and now it will be so much easier <3

**Comment by benesch** [comment]
> Also, thank you for fixing this! Beyond CI auto-detecting panics this has always been frustrating for me when manually inspecting logs and now it will be so much easier <3

For sure! It's amazing to me that any Rust program that generates significant log output is going to run into this. I guess everyone just lives with it?

**Pull request merged**